packages/cli-hooks/src/get-hooks.js

```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env node
```

Comment:
What about framing these hooks as standard javascript modules/functions, but detecting whether they are invoked from the CLI using `require.main`? This would allow us to keep the source to be runnable in node natively, and also allows for running the tests in node.

Consider this file `test.js`:

```
module.exports = function () {
  return {some: 'thing'};
}

if (require.main === module) {
  console.log('main is its own module!');
}
```

Now if I run this file from the command line:

```
➜ node test.js
main is its own module!
```

But if I `require` it within node:

```
➜ node
Welcome to Node.js v18.15.0.
Type ".help" for more information.
> require('./test.js')
[Function (anonymous)]
>
```

I think we can use this pattern to:
- not have to hashbang the env to node at the top of each of the hook
- we can have hooks be wrapped in a function that return objects, making it easier to test. Each test file can invoke the hook function. Hook functions can optionally take arguments, too (many of our hooks accept CLI flags) so testing those can be a pain if you have to frame each test in terms of a process invocation (not to mention that should significantly speed the tests up, since we're not shelling out to additional processes).

So I'd envision the get-hooks implementation to look something like:

```
module.exports = function () {
  return { hooks: {}, whatever: {} };
}

if (require.main === module) {
  console.log(JSON.stringify(module.exports()));
}
```

So we leave the basic, condition-less processing of JSON into strings and outputting to stdout to the `if (require.main === module)` conditional, while any actual hook logic is kept to pure javascript in a simple function.

Reply:
This seems so very much cleaner than the magic of IIEF in scripts! Really appreciate the suggestion and will start hacking on it soon 🙌

Reply:
Attempts to use `require.main` raised errors because these scripts are ES modules instead of common JS (I think). I've found that the following has the same effect:

```js
if (fs.realpathSync(process.argv[1]) === fileURLToPath(import.meta.url)) {
  console.log(JSON.stringify(getHooks())); // eslint-disable-line no-console
}
```

and makes the tests so very much faster! But always open to iterations!

Reply:
As long as it works in both node 18 and 20 I'm cool with it

Reply:
Just checked and this is supported with both `v18.19.0` and `v20.8.0` 🚀